File: MATLAB/tests/test_preprocessing.m
classdef test_preprocessing < matlab.unittest.TestCase
    methods (Test)
        function testLowpassFilter(testCase)
            fs = 100;
            cutoff = 10;
            t = 0:1/fs:10-1/fs;
            data = sin(2 * pi * 5 * t) + sin(2 * pi * 20 * t);
            filtered_data = lowpass_filter(data, cutoff, fs);
            
            testCase.verifyTrue(isnumeric(filtered_data));
            testCase.verifyEqual(size(filtered_data), size(data));
            % Basic check: ensure some attenuation of high frequency component
            % This is a very simple check and might need more sophisticated validation
            testCase.verifyLessThan(std(filtered_data), std(data));
        end
        
        function testPreprocessIteration1(testCase)
            % Dummy data: 20 epochs, 3000 samples per epoch
            eeg_data = randn(20, 3000);
            
            % Create a dummy config struct
            config.CURRENT_ITERATION = 1;
            config.LOW_PASS_FILTER_FREQ = 40;
            
            preprocessed_data = preprocessing_preprocess(eeg_data, config);
            
            testCase.verifyTrue(isnumeric(preprocessed_data));
            testCase.verifyEqual(size(preprocessed_data), size(eeg_data));
        end
        
        function testPreprocessOtherIteration(testCase)
            % Dummy data: 20 epochs, 3000 samples per epoch
            eeg_data = randn(20, 3000);
            
            % Create a dummy config struct for an unhandled iteration
            config.CURRENT_ITERATION = 99;
            config.LOW_PASS_FILTER_FREQ = 40;
            
            preprocessed_data = preprocessing_preprocess(eeg_data, config);
            
            testCase.verifyTrue(isnumeric(preprocessed_data));
            testCase.verifyEqual(size(preprocessed_data), size(eeg_data));
            % For other iterations, it should return raw data, so it should be very similar
            testCase.verifyEqual(preprocessed_data, eeg_data, 'AbsTol', 1e-9); % Check if it's essentially the same
        end
    end
end

File: tests/test_preprocessing.py
import numpy as np
from src.preprocessing import lowpass_filter, preprocess
import config

def test_lowpass_filter():
    fs = 100
    cutoff = 10
    data = np.sin(2 * np.pi * 5 * np.arange(0, 10, 1/fs)) + np.sin(2 * np.pi * 20 * np.arange(0, 10, 1/fs))
    filtered_data = lowpass_filter(data, cutoff, fs)
    assert isinstance(filtered_data, np.ndarray)
    assert filtered_data.shape == data.shape
    # Basic check: ensure some attenuation of high frequency component
    # This is a very simple check and might need more sophisticated validation
    assert np.std(filtered_data) < np.std(data) # Expect some reduction in signal power

def test_preprocess_iteration_1():
    # Dummy data: 20 epochs, 3000 samples per epoch
    eeg_data = np.random.randn(20, 3000)
    config.CURRENT_ITERATION = 1
    preprocessed_data = preprocess(eeg_data, config)
    assert isinstance(preprocessed_data, np.ndarray)
    assert preprocessed_data.shape == eeg_data.shape

def test_preprocess_other_iteration():
    # Dummy data: 20 epochs, 3000 samples per epoch
    eeg_data = np.random.randn(20, 3000)
    config.CURRENT_ITERATION = 99 # An iteration not explicitly handled
    preprocessed_data = preprocess(eeg_data, config)
    assert isinstance(preprocessed_data, np.ndarray)
    assert preprocessed_data.shape == eeg_data.shape
    # For other iterations, it should return raw data, so it should be very similar
    assert np.allclose(preprocessed_data, eeg_data) # Check if it's essentially the same
